### PR TITLE
api: Establish API naming convention.

### PR DESCRIPTION
**Pull request by justinmk**
Since this is essentially "zero cost", given recent deprecation work, I took the opportunity to (I hope) add consistency without too much churn. Also stubbed a section in `doc/develop.txt` with some API naming guidelines.

Changes:

```
old name:                   new name:
--------------------------------------------------
nvim_name_to_color          nvim_get_color_by_name
nvim_get_current_buffer     nvim_get_current_buf
nvim_get_current_window     nvim_get_current_win
nvim_get_buffers            nvim_list_bufs
nvim_get_tabpages           nvim_list_tabpages
nvim_get_windows            nvim_list_wins
nvim_set_current_buffer     nvim_set_current_buf
nvim_set_current_window     nvim_set_current_win
nvim_change_directory       nvim_set_current_dir
nvim_tabpage_get_window     nvim_tabpage_get_win
nvim_tabpage_get_windows    nvim_tabpage_list_wins
nvim_win_get_buffer         nvim_win_get_buf
```

### CODE REVIEW

**Comment by ZyX-I** [comment]
`list_buf` and friends are rather ambigious. It looks like you are listing *one* buffer (i.e. listing something that is “inside” because otherwise this does not make sense), not like you are obtaining a list of all buffers. Such functions need plural, not singular.

**Review by ZyX-I — CHANGES_REQUESTED** [review]
`list_buf` and friends are rather ambigious. It looks like you are listing one buffer (i.e. listing something that is “inside” because otherwise this does not make sense), not like you are obtaining a list of all buffers. Such functions need plural, not singular.

**Comment by ZyX-I** [comment]
Oh, fine crosses.

**Review by ZyX-I — APPROVED** [review]

**Comment by ZyX-I** [comment]
And review message is not deletable or editable, as well as I can’t change “approve” back to “reject” right now. Not sure whether this will change after force-push.

**Comment by justinmk** [comment]
If we use plurals then the "list" verb probably isn't needed , can just say "get_bufs". I'll push that change.

**Comment by ZyX-I** [comment]
list is better for the cases when there is both plural and singular form (e.g. for listing windows in tab pages).

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -711,7 +711,7 @@ Integer nvim_buf_add_highlight(Buffer buffer,
 /// @param line_end End of range of lines to clear (exclusive)
 ///                 or -1 to clear to end of file.
 /// @param[out] err Details of an error that may have occurred
-void nvim_buf_clear_highlight(Buffer buffer,
+void nvim_buf_del_highlight(Buffer buffer,
```
Comment (bfredl) [inline src/nvim/api/buffer.c]: IMHO this is wrong, this does not "delete" a highlight in the way you "delete" a specific key from a dict, rather it clears _all_ highlights of a source from a specified region of the file (typically the entire file).

(Also why is the "add single comment" button greyed out as if you are not supposed to use it? I get two emails every time someone does any "review" operation... Sorry OT)

**Comment by justinmk** [comment]
@bfredl is the difference significant enough ?

**Comment by bfredl** [comment]
It is not the same operation.  del will delete one specified object, or throw an error. `clear_highlight` will silently do nothing, if the file already is clear of highlights, or remove one or many highlights, whatever needed. The point of a convention is surely that the same/analogous operation has the same name, but different operations have different names.

**Comment by justinmk** [comment]
Ok. But I didn't expect to limit "delete" to only one item. Just like "get" may take a query or parameters, and "set" may set many lines etc.

**Comment by jamessan** [comment]
> But I didn't expect to limit "delete" to only one item.

But do you expect "delete" to fail if the specified item doesn't exist?  That's the crux of @bfredl's point, I think.

"clear" means "do whatever is needed, even nothing, to clear all related state".  "delete" means "get rid of these specific items".  If those items don't exist, that should be an error.

**Comment by bfredl** [comment]
but `set_lines` you can still set some _specific_ set of lines. You can't delete one _or_ several specific highlights. `clear_highilght` referers to the intended result, the range/buffer is _clear_ of highlights (highlights may get deleted or may not get deleted, whatever needed) the same way you can clear a dict of all its keys, which is not the same as deleting many specific keys at once.

**Comment by justinmk** [comment]
Ok

**Comment by justinmk** [comment]
Made those changes. Also renamed `change_directory` to `set_current_dir` for parallel form with the other `set_current_*` functions.

**Comment by justinmk** [comment]
UBSAN failure is a unrelated timing issue. If there are no objections I'll merge in the next day.

File: src/nvim/api/vim.c [hunk 2]
```diff
@@ -289,7 +289,7 @@ ArrayOf(String) nvim_list_runtime_paths(void)
 ///
 /// @param dir The new working directory
 /// @param[out] err Details of an error that may have occurred
-void nvim_change_directory(String dir, Error *err)
+void nvim_set_current_dir(String dir, Error *err)
```
Comment (bfredl) [inline src/nvim/api/vim.c]: Can be fixed in a later PR, but API users would expect to also have `nvim_get_current_dir`, by symmetry or something.

**Comment by bfredl** [comment]
LGTM :+1:

**Pull request merged**